### ISSUE TITLE
'Gather Extruded 2D Nodal Parameter' routine much slower with Epetra than Tpetra

### ISSUE THREAD

**Issue by ikalash** [SNLComputation/Albany#523]
In doing performance runs, I discovered that 'Gather Extruded 2D Nodal Parameter' is much slower in Epetra than Tpetra.  This seems to show up for problems with tetrahedral meshes for ALI much more so than hexahedral meshes.  Here are some data:

Epetra:
```
Phalanx: Evaluator 5: [<Residual>] Gather Extruded 2D Nodal Parameter: 21.205 - 28.6921% [23025] {min=20.9137, max=40.8765, std dev=1.55442}
Phalanx: Evaluator 85: [<Jacobian>] Gather Extruded 2D Nodal Parameter: 12.6985 - 19.9394% [13330] {min=12.5126, max=25.1912, std dev=1.01228}
```

Tpetra:
```
Phalanx: Evaluator 5: [<Residual>] Gather Extruded 2D Nodal Parameter: 0.443377 - 2.6535% [23025] {min=0.427363, max=0.693777, std dev=0.0190795}
Phalanx: Evaluator 85: [<Jacobian>] Gather Extruded 2D Nodal Parameter: 0.262684 - 0.8327% [13330] {min=0.252668, max=0.422806, std dev=0.0126421}
```

@bartgol , @mperego : can one of you please have a look at this?  I haven't checked if the issue was happening earlier prior to some of the Thyra refactors - I suspect it's likely not necessary to do that, but I can if you think it would be helpful.

**Comment by ikalash** [comment]
One additional piece of information: seems this evaluator is not called for the hex meshes, only the tet ones.  That explains why I don't see the performance difference b/w Epetra and Tpetra for the hex meshes.

**Comment by mperego** [comment]
Thanks Irina, for pointing it out!

**Comment by ikalash** [comment]
Sure.  Can one of you guys look at it?  Hopefully it's not too hard to fix.

**Comment by bartgol** [comment]
I'm assuming the long one is the Jacobian case, while the shorter one is the Residual. We should probably make sure all our evaluators have the `PHX::typeAsString<EvalT>()` bit int their name, so that we can debug faster.

**Comment by ikalash** [comment]
Sorry I cut that part out by accident.  Actually, the longer one is the Residual.  I've corrected it in the original issue.

**Comment by bartgol** [comment]
Interesting. There's only one possible difference I see: when calling `getLocalElement`, in the Epetra case we perform a check that the input gid fits into a 32bits int (due to limitations in Epetra). However, this check is performed also in other evaluators (notably, the scatter residual one), so I'm reluctant to blame this couple of lines.

I'm at a loss.

**Comment by ikalash** [comment]
I think this issue does have to do with some of the refactors in the spring.  Here is some data from CDash for the  FO_GIS_GisAdjointSensitivityStiffeningBasalFriction problem for the ```Phalanx: Evaluator 4: [<Residual>] Gather Extruded 2D Nodal Parameter``` operation: 0.0507265 - 7.86129% [3696] {min=0.0496671, max=0.0521676, std dev=0.00106272}

3/14, Epetra code stack: 0.0507265 s (http://cdash.sandia.gov/CDash-2-3-0/testDetails.php?test=4393104&build=82442)
3/14, Tpetra code stack: 0.00375801 s (http://cdash.sandia.gov/CDash-2-3-0/testDetails.php?test=4393113&build=82442)
10/2: Epetra code stack: 3.49192 s (http://cdash.sandia.gov/CDash-2-3-0/testDetails.php?test=4393113&build=82442) 
10/2: Tpetra code stack:  0.0156167 

Seems there was about a 10x slowdown with Tpetra compared to a 100 x slowdown with Epetra.

**Comment by bartgol** [comment]
A tiny bit slowdown may be due to the presence of numerous dynamic casts. 10x seems a lot, but I guess this is a relatively small problem, so the hit is more visible. In the numbers you posted at the beginning, the timing was 0.44s for the residual, and I guess that would be closer to the spring timings, due to larger problem sizes (which hide the casts overhead time).

I talked to @mperego and we both can't think of anything in the body of that evaluator that could be to blame. As I said, there's not much in there that is not used already by other evaluators, so this is a bit puzzling.

**Comment by ikalash** [comment]
But we have the dynamic casts for both Epetra and Tpetra, right?  How come the slowdown is 100x for Epetra and only 10x for Tpetra?  Could it be something in Trilinos?  Does any other routine besides PHAL_GatherScalarNodalParameter.hpp come into play for the Gather Extruded nodal field evaluation?

**Comment by bartgol** [comment]
I was referring to the fact that you noticed a 10x slowdown in Tpetra.

Yes, we do casts for both T and E, but the T cast comes first, so E should see roughly twice as many casts. Obviously, that does not explain the slowdown. The Thyra utils we use in that evaluator are very limited, and do not seem to have much overhead (and, as I said, they are used elsewhere, so the drastick difference should be noticed elsewhere too).

As I mentioned at the beginning, the only "big" different between E and T is in `getLocalElement`, where the input GID is checked against the largest 32 bits int (b/c the Epetra stack only works with 32 bits). You can try to comment out those lines, and see if there's a change. I'd be surprised, but I was surprised about the jacobian scatter fix too, so never say never...

**Comment by ikalash** [comment]
OK, I put timers all over this routine 

https://github.com/SNLComputation/Albany/blob/master/src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp#L172-L197

in a Mar. 21 build (57dcb84bc8d46a77a1891d85634933fc41da9dd8 ) and a Mar. 7 build (a63d0afc9a0a8a5c5b67c630c1c5cf08fa98a4a0  ).  Both builds use the same Trilinos.  The getLocalElement call here 

https://github.com/SNLComputation/Albany/blob/master/src/evaluators/gather/PHAL_GatherScalarNodalParameter_Def.hpp#L188 

is the main problem.  Here are times for this routine from Mar. 7 vs. Mar. 21:

Mar. 7: 4ms
Mar. 21: 187ms

Moreover, it appears this routine takes time more often now than in the old version of the code.  What I mean is that when I print the timer in the Mar. 7 version, there are many instances where the routine is reported to take no time; whereas it always takes time in the Mar. 21 version.  If I comment out these lines https://github.com/SNLComputation/Albany/blob/master/src/utility/Albany_ThyraUtils.cpp#L127-L129 per your suggestion @bartgol , the times do not change, so it is not the logic pertaining to 32 vs. 64 bit that is causing the problem.  I'm attaching the relevant file with the timers put in, as well as the output from the runs showing all the relative times.  @bartgol you should be able to use these files to investigate the relative times similar to what I did.

[goodmar7.txt](https://github.com/SNLComputation/Albany/files/3686965/goodmar7.txt)
[badmar21.txt](https://github.com/SNLComputation/Albany/files/3686966/badmar21.txt)
[PHAL_GatherScalarNodalParameter_Def_Mar7.hpp.txt](https://github.com/SNLComputation/Albany/files/3686969/PHAL_GatherScalarNodalParameter_Def_Mar7.hpp.txt)
[PHAL_GatherScalarNodalParameter_Def_Mar21.hpp.txt](https://github.com/SNLComputation/Albany/files/3686970/PHAL_GatherScalarNodalParameter_Def_Mar21.hpp.txt)

**Comment by bartgol** [comment]
@ikalash can you give me the sha of the trilinos commit you used for both those builds?

**Comment by ikalash** [comment]
Sure: c4fbe2ba5c487c1f23caeda3534172140b71f2c5 .  It's from July.  The numbers above are for this problem run on 1 proc: https://github.com/SNLComputation/Albany/blob/master/tests/small/LandIce/FO_GIS/input_fo_gis_analysis_stiffening.yaml .

**Comment by bartgol** [comment]
Did you also have to do a bunch of tweaks to the albany src code, due to a too recent Trilinos version? I'm getting a few compiler errors. Not hard to fix, but I'm just making sure I have what you had.

**Comment by ikalash** [comment]
@bartgol , yes some tweaks were required, mostly to the STK files I believe.  Here are the files modified in my version of the code: 

```
modified:   src/disc/stk/Albany_GenericSTKMeshStruct.cpp
modified:   src/disc/stk/Albany_SideSetSTKMeshStruct.cpp
```

Sorry for not mentioning this before - I had forgotten about these!

**Comment by bartgol** [comment]
Cool, thanks!

**Comment by bartgol** [comment]
Ok, I think the diagnosis is that the methods that require cast to tpetra/epetra must not happen inside loops. Especially the epetra ones, since they can involve checks to the RCP's extra data (due to the lack of a Thyra::EpetraThyraVector, and similar). The Tpetra case sees a lesser overhead, also because it's the first cast to happen (I bet if we swapped T/E order in the casts, Tpetra will see the big hit, I'm going to check this claim), but I still see a bump from 0.17s to 0.95s, so about 5x.

I think this urges us to find a way to have an equivalenf ot a Map object, but petra-agnostic. This should probably be done by resorting to something like Panzer's ConnManager (which, btw, is what I was suggesting in a comment precisely in the GatherScalarExtruded2DNodalParameter's evaluateFields method.

I am thinking the bottleneck is not the dyn cast itself (we do plenty of those in other evaluators, to check E/T, even multiple times per evaluation). The problem must be in the `getEpetraBlockMap` call, when we try to access all the extra data structures. I'm going to check this now.

**Comment by bartgol** [comment]
Confirmed. We were actually doing not the smartest thing anyways in `getEpetraBlockMap`: as explained in the code, there are 3 possible scenarios, and we went ahead and tried to extract two different extra data from the RCP right away. Turns out, the first one was always a hit, so the second one was never really "needed". I isolated the code better (the second extra data is looked for only if the first one is a miss), and the evaluator total time roughly halves, confirming that calling `get_optional_extra_data`/`get_extra_data` is *not* cheap, and should *not* happen inside a loop.

Now I have to think about the best way to overcome this.

**Comment by bartgol** [comment]
My doubt, however, is "why don't we see this slowdown elsewhere"? Or maybe, "are we not realizing there is this slowdown elsewhere"? I'm going to answer the latter in a second...

**Comment by bartgol** [comment]
Ok, there are few places, inside evaluators, where we use `getLocalElements` and `getGlobalElements`. Usually, they are confined scenarios with extruded parameters/fields and DistParamDeriv evaluation type (and sometimes only for transposed derivative). So this explains why we only found this slowdown in this part of the code.

Nevertheless, the slowdown is huge, and demands for a solution. I'll be working on a map-like class (I think Panzer uses the `GlobalIndexer` name for this, but I'm not sure), which should allow one to extract the T/E map once, then reuse it inside the evaluator's loops.

**Comment by bartgol** [comment]
Ok, I implemented an indexer class, which wraps the E/T maps; the indexer is created once at the beginning of the evaluateFields call (we could try to push it to the constructor, if we tried hard, but may not be worth the effort), which means only one map extraction per evaluateFields call.

I tried it in the `GatherScalarExtruded2DNodalParameter` class, and the evaluation time for the Residual goes down from 20s to 0.3s. I'm going to put this utility class in all the other classes that use `getLocalElement` or `getGlobalElement`, so we won't see this kind of slowdowns again in future runs (finger crossed). Then I'll submit a PR.
